src/lib.rs

```diff
@@ -1,7 +1,168 @@
+    type M: GuestMemory<'a>;
+
+    /// Set the current process as the owner of the vhost backend device.
+    /// This must be run before any other vhost commands.
+    fn set_owner(&self) -> Result<()>;
+
+    /// Get a bitmask of supported virtio/vhost features.
+    fn get_features(&self) -> Result<u64>;
+
+    /// Inform the vhost subsystem which features to enable.
+    /// This should be a subset of supported features from get_features().
+    ///
+    /// # Arguments
+    /// * `features` - Bitmask of features to set.
+    fn set_features(&self, features: u64) -> Result<()>;
+
+    /// Set the guest memory mappings for vhost to use.
+    fn set_mem_table(&self, mem: &'a Self::M) -> Result<()>;
+
+    /// Set the number of descriptors in the vring.
+    ///
+    /// # Arguments
+    /// * `queue_index` - Index of the queue to set descriptor count for.
+    /// * `num` - Number of descriptors in the queue.
+    fn set_vring_num(&self, queue_index: usize, num: u16) -> Result<()>;
+
+    /// Set the addresses for a given vring.
+    ///
+    /// # Arguments
+    /// * `queue_max_size` - Maximum queue size supported by the device.
+    /// * `queue_size` - Actual queue size negotiated by the driver.
+    /// * `queue_index` - Index of the queue to set addresses for.
+    /// * `flags` - Bitmask of vring flags.
+    /// * `desc_table_addr` - Descriptor table address.
+    /// * `used_ring_addr` - Used ring buffer address.
+    /// * `avail_ring_addr` - Available ring buffer address.
+    /// * `log_addr` - Optional address for logging.
+    #[allow(clippy::too_many_arguments)]
+    fn set_vring_addr(
+        &self,
+        queue_max_size: u16,
+        queue_size: u16,
+        queue_index: usize,
+        flags: u32,
+        desc_table_addr: GuestAddress,
+        used_ring_addr: GuestAddress,
+        avail_ring_addr: GuestAddress,
+        log_addr: Option<GuestAddress>,
+    ) -> Result<()>;
+
+    /// Set the first index to look for available descriptors.
+    ///
+    /// # Arguments
+    /// * `queue_index` - Index of the queue to modify.
+    /// * `num` - Index where available descriptors start.
+    fn set_vring_base(&self, queue_index: usize, num: u16) -> Result<()>;
+
+    /// Set the eventfd to trigger when buffers have been used by the host.
+    ///
+    /// # Arguments
+    /// * `queue_index` - Index of the queue to modify.
+    /// * `fd` - EventFd to trigger.
+    fn set_vring_call(&self, queue_index: usize, fd: &EventFd) -> Result<()>;
+
+    /// Set the eventfd that will be signaled by the guest when buffers are
+    /// available for the host to process.
+    ///
+    /// # Arguments
+    /// * `queue_index` - Index of the queue to modify.
+    /// * `fd` - EventFd that will be signaled from guest.
+    fn set_vring_kick(&self, queue_index: usize, fd: &EventFd) -> Result<()>;
+}
+
+/// An interface for controlling vsock vHost backend devices.
+#[cfg(feature = "vhost-vsock")]
+pub trait VhostVsock<'a>: VhostDevice<'a> {
```

Comment:
We could define this one in its own `vsock.rs` file.

And also, why do you need this to be a `trait`? Do you expect multiple implementation of this?

Reply:
Sure, I will move it to vsock.rs.
The existing vsock implementation is used a sample, because net and blk is not ready yet.